sklearn/preprocessing/imputation.py

```diff
@@ -366,13 +396,66 @@ def transform(self, X):
 
             mask = _get_mask(X, self.missing_values)
             n_missing = np.sum(mask, axis=self.axis)
-            values = np.repeat(valid_statistics, n_missing)
 
-            if self.axis == 0:
-                coordinates = np.where(mask.transpose())[::-1]
+            if self.strategy == 'knn':
+                if self.axis == 1:
+                    X = X.transpose()
+                    mask = mask.transpose()
+                    statistics = statistics.transpose()
+
+                batch_size = 10  # set batch size for block query
+                if False:
+                    missing_index = np.where(mask.any(1))[0]
+                    #@jnothman 's method
```

Comment:
can you please refactor this method to not slice twice and then benchmark again? I feel it is to complicated currently (or I'm missing something).